File: Graph/MST/1584. Min Cost to Connect All Points/1584.java
class Solution {
    public int minCostConnectPoints(int[][] points) {
        int N = points.length;

        Set<Integer> rest = new HashSet<>();
        for (int i = 0; i < N; i++) {
            rest.add(i);
        }

        int[] shortestD = new int[N];
        Arrays.fill(shortestD, Integer.MAX_VALUE);

        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        minHeap.offer(new int[]{0, 0});

        int ans = 0;

        while(!rest.isEmpty()){
            int[] pair = minHeap.poll();
            int d = pair[0];
            int u = pair[1];
            if (!rest.contains(u)) {continue;}
            rest.remove(u);
            ans += d;
            int x1 = points[u][0];
            int y1 = points[u][1];
            for (int v : rest) {
                int x2 = points[v][0];
                int y2 = points[v][1];
                d = Math.abs(x1 - x2) + Math.abs(y1 - y2);
                if (d < shortestD[v]) {
                    shortestD[v] = d;
                    minHeap.add(new int[]{d, v});
                }
            }
        }
        return ans;
    }
}

File: Graph/MST/1584. Min Cost to Connect All Points/1584.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        N = len(points)
        rest = {i for i in range(N)}
        shortest_d = [math.inf] * N
        heap = [[0,0]]
        ans = 0
        while rest:
            d, u = heapq.heappop(heap)
            if u not in rest: continue
            rest.remove(u)
            ans += d
            x1, y1 = points[u]
            for v in rest:
                x2, y2 = points[v]
                d = abs(x1 - x2) + abs(y1 - y2)
                if d < shortest_d[v]:
                    shortest_d[v] = d
                    heapq.heappush(heap, [d, v])
        
        return ans
